File: app-sdk/src/main/java/code/sdk/download/DownloadTask.java
package code.sdk.download;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;

import io.reactivex.rxjava3.annotations.NonNull;
import io.reactivex.rxjava3.core.Observer;
import io.reactivex.rxjava3.disposables.Disposable;
import okhttp3.ResponseBody;


public class DownloadTask {
    public static final String TAG = DownloadTask.class.getSimpleName();

    private static volatile DownloadTask mInstance;

    public static DownloadTask getInstance() {
        if (mInstance == null) {
            synchronized (DownloadTask.class) {
                if (mInstance == null) {
                    mInstance = new DownloadTask();
                }
            }
        }
        return mInstance;
    }

    private DownloadTask() {

    }

    /**
     * @param url      下载连接
     * @param saveDir  储存下载文件的SDCard目录
     * @param listener 下载监听
     */
    public void download(final String url, final String saveDir, final OnDownloadListener listener) {
        DownloadClient instance = DownloadClient.getInstance();
        instance.getApi().download(url)
                .map(responseBody -> saveFile(responseBody, saveDir, listener))
                .compose(instance.ioSchedulers())
                .subscribe(new Observer<File>() {
                    @Override
                    public void onSubscribe(@NonNull Disposable d) {
                    }

                    @Override
                    public void onNext(@NonNull File file) {
                        if (file == null) {
                            listener.onDownloadFailed();
                        }else{
                            listener.onDownloadSuccess(file);
                        }
                    }

                    @Override
                    public void onError(@NonNull Throwable e) {
                        listener.onDownloadFailed();
                    }

                    @Override
                    public void onComplete() {}

                });
    }


    private File saveFile(ResponseBody responseBody, String saveDir, OnDownloadListener listener) {
        File img = null;
        InputStream is = null;
        byte[] buf = new byte[2048];
        int len = 0;
        FileOutputStream fos = null;
        try {
            is = responseBody.byteStream();
            long total = responseBody.contentLength();
            File saveFile = new File(saveDir, System.currentTimeMillis() + ".jpg");
            fos = new FileOutputStream(saveFile);
            long sum = 0;
            while ((len = is.read(buf)) != -1) {
                fos.write(buf, 0, len);
                sum += len;
                int progress = (int) (sum * 1.0f / total * 100);
                listener.onDownloading(progress);
            }
            fos.flush();
            img = saveFile;
        } catch (Exception e) {

        } finally {
            try {
                if (is != null)
                    is.close();
            } catch (IOException e) {
            }
            try {
                if (fos != null)
                    fos.close();
            } catch (IOException e) {
            }
        }
        return img;
    }

    /**
     * @param url
     * @return 从下载连接中解析出文件名
     */
    public String getNameFromUrl(String url) {
        return url.substring(url.lastIndexOf("/") + 1);
    }

    public interface OnDownloadListener {
        /**
         * 下载成功
         */
        void onDownloadSuccess(File saveFile);

        /**
         * @param progress 下载进度
         */
        void onDownloading(int progress);

        /**
         * 下载失败
         */
        void onDownloadFailed();
    }
}

File: app-sdk/src/main/java/code/sdk/download/DownloadTask.kt
package code.sdk.download

import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.disposables.Disposable
import okhttp3.ResponseBody
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream

class DownloadTask private constructor() {
    /**
     * @param url      下载连接
     * @param saveDir  储存下载文件的SDCard目录
     * @param listener 下载监听
     */
    fun download(url: String?, saveDir: String, listener: OnDownloadListener) {
        val instance = DownloadClient.mInstance
        instance.getApi().download(url)
            .map { t -> saveFile(t, saveDir, listener) }
            .compose(instance.ioSchedulers())
            .subscribe(object : Observer<File> {
                override fun onSubscribe(d: Disposable) {}
                override fun onNext(file: File) {
                    if (!file.exists()) {
                        listener.onDownloadFailed()
                    } else {
                        listener.onDownloadSuccess(file)
                    }
                }

                override fun onError(e: Throwable) {
                    listener.onDownloadFailed()
                }

                override fun onComplete() {}
            })
    }

    private fun saveFile(
        responseBody: ResponseBody,
        saveDir: String,
        listener: OnDownloadListener
    ): File {
        val img = File(saveDir, System.currentTimeMillis().toString() + ".jpg")
        var inputStream: InputStream? = null
        val buf = ByteArray(2048)
        var len: Int
        var fos: FileOutputStream? = null
        try {
            inputStream = responseBody.byteStream()
            val total = responseBody.contentLength()

            fos = FileOutputStream(img)
            var sum: Long = 0
            while (inputStream.read(buf).also { len = it } != -1) {
                fos.write(buf, 0, len)
                sum += len.toLong()
                val progress = (sum * 1.0f / total * 100).toInt()
                listener.onDownloading(progress)
            }
            fos.flush()
        } catch (_: Exception) {
        } finally {
            try {
                inputStream?.close()
            } catch (_: IOException) {
            }
            try {
                fos?.close()
            } catch (_: IOException) {
            }
        }
        return img
    }

    /**
     * @param url
     * @return 从下载连接中解析出文件名
     */
    fun getNameFromUrl(url: String): String {
        return url.substring(url.lastIndexOf("/") + 1)
    }

    interface OnDownloadListener {
        /**
         * 下载成功
         */
        fun onDownloadSuccess(saveFile: File)

        /**
         * @param progress 下载进度
         */
        fun onDownloading(progress: Int)

        /**
         * 下载失败
         */
        fun onDownloadFailed()
    }

    companion object {
       private val TAG = DownloadTask::class.java.simpleName

        val mInstance by lazy { DownloadTask() }
    }
}
